### ISSUE TITLE
Discrepancies in SparseAttentionMeansim Results

### ISSUE THREAD

**Issue by tobygh** [thu-ml/SpargeAttn#47]
Dear authors,

Thank you so much for your hard work on this project. It has been incredibly helpful to me, and I'm excited about the potential of SparseAttentionMeansim in my research/work.

However, I've encountered an issue where the results of SparseAttentionMeansim do not align with the results of a standard attention mechanism. This discrepancy is significant and seems to be beyond the expected minor differences due to quantization.

**Modifications Made**
To troubleshoot this issue, I made the following modifications:

1. Set final_map[:] = 1 in utils.py at line 292:
- [Link to ](https://github.com/thu-ml/SpargeAttn/blob/main/spas_sage_attn/utils.py#L292)
- This modification ensures that all blocks will be calculated, ignoring simthreshold and cdfthreshold, as far as I understand.
2. Replace the kernel in core.py at line 40:
- [Link to ](https://github.com/thu-ml/SpargeAttn/blob/main/spas_sage_attn/core.py#L40)
- Original line:
`qattn.qk_int8_sv_f16_accum_f16_block_sparse_attn_inst_buf_with_pv_threshold(q_int8, k_int8, v, o, lut, valid_block_num, pvthreshd, q_scale, k_scale, 1, _is_causal, 1, scale, 0)
`
- Modified line:
`qattn.qk_int8_sv_f16_accum_f16_block_sparse_attn_inst_buf(q_int8, k_int8, v, o, lut, valid_block_num, q_scale, k_scale, 1, _is_causal, 1, scale)`
- This modification ignores PV threshold, as far as I understand.
**Expected Behavior**
With these modifications, I expected that SparseAttentionMeansim would not skip any calculations, and the differences between the results of SparseAttentionMeansim and standard attention should only be minor (on the order of 1e-6) due to the quantization of Q and K.

**Observed Behavior**
However, the observed differences are much larger than expected. The results of SparseAttentionMeansim still show significant discrepancies compared to the standard attention mechanism.

**Additional Information**
- **Environment**:
  - Hardware: A100, cuda-12.0
  - Software: torch==2.2.0+cu121, gcc=8.2.0
- **Reproducibility**:
  - I have consistently observed this issue across multiple runs with different input data.
  - I can provide a minimal reproducible example if needed.

**Questions**
1. Is there any known issue with the SparseAttentionMeansim implementation that might cause such significant discrepancies?
2. Are there any additional parameters or configurations that I should be aware of to ensure accurate results?
3. Could you please provide some guidance on how to further diagnose and resolve this issue?

Thank you very much for your time and assistance. I believe resolving this issue could significantly improve the usability and accuracy of SparseAttentionMeansim.

Best regards,
Toby

**Comment by tobygh** [comment]
@Xiang-cd Sure! I can provide the metrics generated.

When V is an Identity Matrix
- metric: `Cossim: 0.999900, L1: 0.010600, RMSE:0.000200`
- diff: max 0.003173828125 mean 8.285045623779297e-05

When V is random Matrix
- metric: `Cossim: 0.999900, L1: 0.012900, RMSE:0.001800`
- diff: max 0.01220703125 mean 0.0014104843139648438

You could find the detailed output in [log_debug.txt](https://github.com/user-attachments/files/19929559/log_debug.txt)

Here is Code

```Python3
import torch
import math
import torch.nn as nn
from spas_sage_attn.autotune import SparseAttentionMeansim

import numpy as np
np.random.seed(123)
torch.manual_seed(123)

import torch.nn.functional as F
def tune(new_attention, q, k, v) :
    new_attention(q, k, v, tune_mode = True)

def naive_attn(q, k, v):
    return F.scaled_dot_product_attention(q, k, v)

def compare_attn(q, k, v, new_attention : SparseAttentionMeansim) :
    ans = naive_attn(q, k, v)
    pred = new_attention(q, k, v, tune_mode = False)
    diff = torch.abs(pred - ans)
    print("diff max {} mean {}".format(torch.max(torch.flatten(diff)), torch.mean(torch.flatten(diff))))

def case1():
    head_num = 1
    # bs, head_num, seq_len, d_k
    # seq_len >= 128
    device = "cuda"

    hh = 128
    q = torch.randn(1, head_num, 128, hh, device=device, dtype=torch.float16)
    k = torch.randn(1, head_num, 128, hh, device=device, dtype=torch.float16)

    # 构造一个v来探查P矩阵
    v = torch.zeros(1, head_num, 128, hh, device=device, dtype=torch.float16)
    for i in range(hh): 
        v[0, 0, i, i] = 1
    new_attention = SparseAttentionMeansim(l1=0.06, pv_l1=0.065)
    tune(new_attention, q, k, v)
    compare_attn(q, k, v, new_attention)

    # 对比最终结果
    v = torch.randn(1, head_num, 128, hh, device=device, dtype=torch.float16)
    new_attention = SparseAttentionMeansim(l1=0.06, pv_l1=0.065)
    tune(new_attention, q, k, v)
    compare_attn(q, k, v, new_attention)


if __name__ == "__main__":
    case1()

```
